File: frontend/src/api/planner.ts
import { api } from "./client";
import type { DayPlan, Task, WeekPlan } from "@/types/planner";

export interface TaskCreatePayload {
  task_type: "test" | "review" | "study";
  title: string;
  source_type?: string;
  source_id?: number | null;
  node_id?: number | null;
  question_count?: number | null;
  sequence_from?: number | null;
  sequence_to?: number | null;
  parity?: "odd" | "even" | "any" | null;
  priority?: number;
  estimated_minutes?: number;
  due_at?: string | null;
  recommendation_reason?: string | null;
}

export function createTask(payload: TaskCreatePayload): Promise<Task> {
  return api.post<Task>("/tasks", payload);
}

export function patchTask(
  taskId: number,
  payload: Record<string, unknown>,
): Promise<Task> {
  return api.patch<Task>(`/tasks/${taskId}`, payload);
}

export function completeTask(taskId: number): Promise<Task> {
  return api.post<Task>(`/tasks/${taskId}/complete`);
}

export function putPlacements(
  placements: Array<{ task_id: number; date: string; position?: number }>,
  dates?: string[],
): Promise<DayPlan[]> {
  return api.put<DayPlan[]>("/planner/placements", { placements, dates: dates ?? [] });
}

export function fetchDay(day: string): Promise<DayPlan> {
  return api.get<DayPlan>(`/planner/day/${day}`);
}

export function fetchWeek(week: string): Promise<WeekPlan> {
  return api.get<WeekPlan>(`/planner/week/${week}`);
}

export function setOverride(
  day: string,
  isSchoolDay: boolean,
  reason?: string,
): Promise<{ date: string; is_school_day: boolean; reason: string | null }> {
  return api.post("/school-day-overrides", {
    date: day,
    is_school_day: isSchoolDay,
    reason: reason ?? null,
  });
}

File: backend/app/api/v1/planner.py
"""Planner routes (spec 14). Routes only."""

from datetime import date

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.api.deps import get_current_user_id
from app.db import get_db
from app.errors import AppError
from app.schemas.planner import (
    DayPlanOut,
    OverrideIn,
    OverrideOut,
    PlacementsPut,
    TaskCreate,
    TaskOut,
    TaskPatch,
    WeekPlanOut,
)
from app.services import planner as service

router = APIRouter(tags=["planner"])


@router.post("/tasks", response_model=TaskOut)
def create_task(
    payload: TaskCreate,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> TaskOut:
    return service.create_task(db, user_id=user_id, payload=payload)


@router.patch("/tasks/{task_id}", response_model=TaskOut)
def patch_task(
    task_id: int,
    payload: TaskPatch,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> TaskOut:
    return service.patch_task(db, user_id=user_id, task_id=task_id, payload=payload)


@router.post("/tasks/{task_id}/complete", response_model=TaskOut)
def complete_task(
    task_id: int,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> TaskOut:
    return service.complete_task(db, user_id=user_id, task_id=task_id)


@router.put("/planner/placements", response_model=list[DayPlanOut])
def put_placements(
    payload: PlacementsPut,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> list[DayPlanOut]:
    return service.put_placements(
        db, user_id=user_id, items=payload.placements, extra_dates=payload.dates)


@router.get("/planner/day/{day}", response_model=DayPlanOut)
def get_day(
    day: str,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> DayPlanOut:
    try:
        parsed = date.fromisoformat(day)
    except ValueError:
        raise AppError("invalid_date", "قالب تاریخ باید YYYY-MM-DD باشد.", status_code=422)
    return service.day_plan(db, user_id=user_id, day=parsed)


@router.get("/planner/week/{week}", response_model=WeekPlanOut)
def get_week(
    week: str,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> WeekPlanOut:
    return service.week_plan(db, user_id=user_id, week=week)


# Owned by capacity (Phase 5); lives with Academic in the contract.
@router.post("/school-day-overrides", response_model=OverrideOut)
def set_override(
    payload: OverrideIn,
    db: Session = Depends(get_db),
    user_id: int = Depends(get_current_user_id),
) -> OverrideOut:
    return service.set_override(db, user_id=user_id, payload=payload)
